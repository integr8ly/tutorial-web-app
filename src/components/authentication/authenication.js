import React from 'react';
import PropTypes from 'prop-types';
import { noop, Button, Card, Form, Grid } from 'patternfly-react';
import { connect, reduxActions } from '../../redux';
import apiTypes from '../../constants/apiConstants';
import { fieldValidation } from '../formField/formField';
import { OC_MODE } from '../../common/helpers';
import titleImgBrand from '../../img/login-reversed.svg';

// todo: use patternfly-react LoginPage comonent

class Authentication extends React.Component {
  state = {
    email: '',
    emailError: null,
    password: '',
    passwordError: null,
    remember: false,
    formTouched: false,
    formValid: false
  };

  componentDidMount() {
    const { session, checkUser, storeData } = this.props;

    if (!session.authorized) {
      checkUser();
      storeData();
    }
  }

  onChangeEmail = event => {
    const { value } = event.target;
    const errorMessage = fieldValidation.isEmpty(value) ? 'Email must be valid' : '';

    this.setState(
      {
        email: value,
        emailError: errorMessage,
        formTouched: true
      },
      () => this.isFormValid()
    );
  };

  onChangePassword = event => {
    const { value } = event.target;
    const errorMessage = fieldValidation.isEmpty(value) ? 'Password must be valid' : '';

    this.setState(
      {
        password: value,
        passwordError: errorMessage,
        formTouched: true
      },
      () => this.isFormValid()
    );
  };

  onChangeRemember = event => {
    const { checked } = event.target;
    const { removeStoredData } = this.props;

    if (!checked) {
      removeStoredData().then(() =>
        this.setState({
          remember: checked
        })
      );
    } else {
      this.setState({
        remember: checked
      });
    }
  };

  onLogin = event => {
    const { email, password, remember, formValid } = this.state;
    const { checkUser, loginUser, removeStoredData, storeData } = this.props;

    event.preventDefault();

    if (formValid) {
      this.setState(
        {
          formTouched: false,
          password: '',
          passwordError: null
        },
        () =>
          loginUser({
            [apiTypes.API_SUBMIT_AUTH_USERNAME]: email,
            [apiTypes.API_SUBMIT_AUTH_PASSWORD]: password
          }).then(() => {
            if (remember) {
              storeData({ email });
            } else {
              this.setState(
                {
                  email: '',
                  emailError: null
                },
                () => removeStoredData()
              );
            }

            return checkUser();
          })
      );
    }
  };

  static getDerivedStateFromProps(props, state) {
    let updateInitialState = null;

    if (OC_MODE) {
      updateInitialState = {
        email: (props.session.remember && props.session.storedEmail) || state.email,
        password: state.password,
        emailError: '',
        passwordError: '',
        formValid: true
      };
    } else if (!state.formTouched && props.session.remember && props.session.storedEmail) {
      updateInitialState = {
        email: props.session.storedEmail,
        emailError: '',
        remember: props.session.remember
      };
    }

    return updateInitialState;
  }

  isFormValid() {
    const { emailError, passwordError } = this.state;
    const formValid = emailError === '' && passwordError === '';

    this.setState({
      formValid
    });
  }

  static renderLoading(message = 'Loading...') {
    return (
      <Card className="app-login-loading-card">
        <Card.Body>
          <div className="spinner spinner-xl" />
          <div className="text-center">{message}</div>
        </Card.Body>
      </Card>
    );
  }

  renderLogin() {
    const { email, emailError, formTouched, password, passwordError, remember } = this.state;
    const { session } = this.props;

    return (
      <Card className="app-login-card">
        <header className="login-pf-header">
          <select className="selectpicker">
            <option>English</option>
          </select>
          <h1>Log In to Your Account</h1>
        </header>
        <Card.Body>
          <Form method="post" autoComplete={remember ? 'on' : 'off'} onSubmit={this.onLogin}>
            <div className="app-login-card-error help-block" aria-live="polite">
              {(!formTouched && session.error && session.loginFailed) ||
              (emailError !== '' && emailError !== null) ||
              (passwordError !== '' && passwordError !== null)
                ? 'Email address or password is incorrect.'
                : null}
            </div>
            <Form.FormGroup controlId="email">
              <Form.ControlLabel srOnly>Email address</Form.ControlLabel>
              <Form.FormControl
                bsSize="lg"
                type="email"
                value={email}
                placeholder="Email address"
                required
                name="email"
                onChange={this.onChangeEmail}
              />
            </Form.FormGroup>
            <Form.FormGroup controlId="password">
              <Form.ControlLabel srOnly>Password</Form.ControlLabel>
              <Form.FormControl
                bsSize="lg"
                type="password"
                value={password}
                placeholder="Password"
                required
                name="password"
                onChange={this.onChangePassword}
              />
            </Form.FormGroup>
            <Form.FormGroup controlId="remember" className="login-pf-settings app-login-settings">
              <Form.Checkbox
                name="remember"
                checked={remember}
                inline
                className="checkbox-label"
                onChange={this.onChangeRemember}
              >
                Remember email address
              </Form.Checkbox>
              <Button bsStyle="link" className="sr-only">
                Forgot password?
              </Button>
            </Form.FormGroup>
            <Button type="submit" bsStyle="primary" bsSize="large" className="btn-block">
              Log In
            </Button>
          </Form>
        </Card.Body>
      </Card>
    );
  }

  render() {
    const { children, session } = this.props;

    if (session.authorized) {
      return children;
    }

    return (
      <div className="login-pf app-login fadein">
        <div className="login-pf-page app-login-body">
          <div className="container-fluid">
            <Grid.Row>
              <Grid.Col sm={8} smOffset={2} md={6} mdOffset={3} lg={6} lgOffset={3}>
                <header className="login-pf-page-header">
                  <img className="login-pf-brand" src={titleImgBrand} alt="PatternFly" />
                </header>
                <Grid.Row>
                  <Grid.Col sm={10} smOffset={1} md={8} mdOffset={2} lg={8} lgOffset={2}>
                    {!session.pending && this.renderLogin()}
                    {session.pending && Authentication.renderLoading()}
                  </Grid.Col>
                </Grid.Row>
              </Grid.Col>
            </Grid.Row>
          </div>
        </div>
      </div>
    );
  }
}

Authentication.propTypes = {
  checkUser: PropTypes.func,
  children: PropTypes.node.isRequired,
  loginUser: PropTypes.func,
  removeStoredData: PropTypes.func,
  session: PropTypes.shape({
    error: PropTypes.bool,
    loginFailed: PropTypes.bool,
    authorized: PropTypes.bool,
    pending: PropTypes.bool,
    storedEmail: PropTypes.string,
    remember: PropTypes.bool
  }),
  storeData: PropTypes.func
};

Authentication.defaultProps = {
  checkUser: noop,
  loginUser: noop,
  removeStoredData: noop,
  session: {},
  storeData: noop
};

const mapDispatchToProps = dispatch => ({
  checkUser: () => dispatch(reduxActions.user.checkUser()),
  loginUser: data => dispatch(reduxActions.user.loginUser(data)),
  storeData: data => dispatch(reduxActions.user.storeData(data)),
  removeStoredData: () => dispatch(reduxActions.user.removeStoredData())
});

const mapStateToProps = state => ({ session: state.user.session });

const ConnectedAuthentication = connect(
  mapStateToProps,
  mapDispatchToProps
)(Authentication);

export { ConnectedAuthentication as default, ConnectedAuthentication, Authentication };
