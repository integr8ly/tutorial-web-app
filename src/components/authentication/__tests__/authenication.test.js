import React from 'react';
import configureMockStore from 'redux-mock-store';
import { shallow, mount } from 'enzyme';
import { ConnectedAuthentication, Authentication } from '../authenication';

describe('Authentication Component', () => {
  const generateEmptyStore = (obj = {}) => configureMockStore()(obj);

  it('should render a basic component with login form', () => {
    const store = generateEmptyStore({ user: { session: { error: false, loginFailed: false, pending: false } } });
    const component = shallow(
      <ConnectedAuthentication>
        <span className="test">lorem</span>
      </ConnectedAuthentication>,
      { context: { store } }
    );

    expect(component).toMatchSnapshot('connected');
  });

  it('should render a basic component without login form', () => {
    const props = {
      session: {
        authorized: true
      }
    };

    const component = mount(
      <Authentication {...props}>
        <span className="test">lorem</span>
      </Authentication>
    );

    expect(component.render()).toMatchSnapshot('post-authorization');
  });

  it('should have specific events defined', () => {
    const checkUser = jest.fn();
    const props = {
      checkUser
    };
    const component = mount(
      <Authentication {...props}>
        <span className="test">lorem</span>
      </Authentication>
    );
    const componentInstance = component.instance();

    expect(componentInstance.onChangeEmail).toBeDefined();
    expect(componentInstance.onChangePassword).toBeDefined();
    expect(componentInstance.onChangeRemember).toBeDefined();
    expect(componentInstance.onLogin).toBeDefined();

    expect(checkUser).toHaveBeenCalled();

    component.find('input[name="email"]').simulate('change', { target: { value: '' } });
    component.find('input[name="password"]').simulate('change', { target: { value: '123' } });
    component.find('input[name="remember"]').simulate('change', { target: { checked: true } });
    expect(componentInstance.state).toMatchSnapshot('expected state');
  });
});
