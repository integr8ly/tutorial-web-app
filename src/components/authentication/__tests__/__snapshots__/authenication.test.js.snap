// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authentication Component should have specific events defined: expected state 1`] = `
Object {
  "email": "",
  "emailError": "Email must be valid",
  "formTouched": true,
  "formValid": false,
  "password": "123",
  "passwordError": "",
  "remember": true,
}
`;

exports[`Authentication Component should render a basic component with login form: connected 1`] = `
<Authentication
  checkUser={[Function]}
  loginUser={[Function]}
  removeStoredData={[Function]}
  session={
    Object {
      "error": false,
      "loginFailed": false,
      "pending": false,
    }
  }
  storeData={[Function]}
>
  <span
    className="test"
  >
    lorem
  </span>
</Authentication>
`;

exports[`Authentication Component should render a basic component without login form: post-authorization 1`] = `
<span
  class="test"
>
  lorem
</span>
`;
